src/librustc_mir/borrow_check/nll/constraint_generation.rs

```diff
@@ -75,14 +84,27 @@ impl<'cx, 'gcx, 'tcx> ConstraintGeneration<'cx, 'gcx, 'tcx> {
                     }
                 });
 
-            self.liveness
-                .drop
-                .simulate_block(self.mir, bb, |location, live_locals| {
-                    for live_local in live_locals.iter() {
+            let mut all_live_locals: Vec<(Location, Vec<Local>)> = vec![];
+            self.liveness.drop.simulate_block(self.mir, bb, |location, live_locals| {
+                all_live_locals.push((location, live_locals.iter().collect()));
+            });
+
+            self.flow_inits.reset_to_entry_of(bb);
+            for index in 0 .. self.mir.basic_blocks()[bb].statements.len() {
```

Comment:
The `statement` value isn't used by `reconstruct_statement_effect`, so I just iterate over the number of statements. Is that still useful as a sanity check, or should we just use the data in `all_live_locals`?

Reply:
Ah, I see, let's just re-use the data. Note that there should be one final entry to pop, corresponding to the `terminator`. If we just collect this into a `while let Some((location, live_locals)) = all_live_locals.pop() { .. }` loop, as you suggest, then we can cleanly combine those two things. We just need to be careful not to invoke `reconstruct_statement_effect` when we are on the last item but rather `reconstruct_terminator_effect` (or nothing at all, since we don't need the final result I don't think).